prov/shm/src/smr_progress.c

```diff
@@ -202,8 +217,93 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
 	return -ret;
 }
 
+static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
+				 struct iovec *iov, size_t iov_count,
+				 size_t *total_len)
+{
+	char shm_name[NAME_MAX];
+	char peer_name[NAME_MAX];
+	void *mapped_ptr;
+	int peer_id, fd, ret, num;
+
+	peer_id = (int) cmd->msg.hdr.addr;
+
+	sprintf(peer_name, "%s", ep->region->map->peers[peer_id].peer.name);
+	num = sprintf(shm_name, "%s_%s_%ld", peer_name, ep->name,
+		      cmd->msg.hdr.msg_id);
+	if (num < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "generating shm file name failed\n");
+		return errno;
+	}
+	assert(num <= NAME_MAX);
+
+	fd = shm_open(shm_name, O_RDWR, S_IRUSR | S_IWUSR);
+	if (fd < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "shm_open error\n");
+		return errno;
+	}
+
+	mapped_ptr = mmap(NULL, cmd->msg.hdr.size, PROT_READ | PROT_WRITE,
+			  MAP_SHARED, fd, 0);
+	if (mapped_ptr == MAP_FAILED) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "mmap error %s\n", strerror(errno));
+		return errno;
+	}
```

Comment:
On failure here, do we still need to shm_unlink?

Reply:
Yeah, updated, thanks.

Reply:
-errno and cleanup

Reply:
Done.

> > + if (mr_mode & FI_MR_VIRT_ADDR) + cur->domain_attr->mr_mode = FI_MR_VIRT_ADDR; My understanding is that if app requires FI_MR_VIRT_ADDR via hints, and shm provider supports FI_MR_VIRT_ADDR, the returned info should have FI_MR_VIRT_ADDR bit, regardless of CMA capability.
> FI_MR_VIRT_ADDR is a requirement *from the provider*, not the app. The app indicates if it can support that requirement. The provider toggles the bit on/off if it is needed.

Sorry for the confusion. Let me put my question in this way:
suppose that app puts FI_MR_VIRT_ADDR bit in hints, and the app calls fi_getinfo. 
Is FI_MR_VIRT_ADDR bit supposed to be returned from shm provider?

The current logic (without my change) is: 
- if CMA available, rma fast path will be enabled, FI_MR_VIRT_ADDR will be set and returned.
- if no CMA, rma fast path will be disabled, thus NO FI_MR_VIRT_ADDR bit will be set and retuned.
Is such behavior expected?